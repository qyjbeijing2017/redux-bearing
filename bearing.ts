import { combineReducers, createStore, Reducer, Store } from "redux";
import { BearingState } from "./bearing-state";
import { composeWithDevTools } from 'redux-devtools-extension'

interface BreaingStartOption {
    [key: string]: BreaingStartOption | { new(pos: string): BearingStates<any> };
}

type BearingStates<T extends BreaingStartOption> = {
    [P in keyof T]: T[P] extends { new(pos?: string): any } ? InstanceType<T[P]> : T[P] extends BreaingStartOption ? BearingStates<T[P]> : null;
}

export const combinedReducer2SamePoint = (reducers: (Reducer | undefined)[]) => {
    return (state: any, action: any) => {
        let stateNew = state;
        for (const reducer of reducers) {
            if (reducer)
                stateNew = { ...stateNew, ...reducer(stateNew, action) };
        }
        return stateNew;
    }
}

export default class Bearing<T extends BreaingStartOption> {
    private readonly _store: Store;
    private readonly _states: BearingStates<T>
    constructor(startType: T, reducer?: Reducer) {
        this._states = this.createStates(startType);
        const myReducer = this.reducer(this._states);
        this._store = createStore(combinedReducer2SamePoint([myReducer as any, reducer]), composeWithDevTools());
        this.foreachStates(this._states, (state) => state.store = this._store);
    }

    get store() {
        return this._store;
    }

    get states(): BearingStates<T> {
        return this._states;
    }

    get storeStates(){
        return this.store.getState();
    }

    release = () => { this.foreachStates(this._states, state => state.release()) }

    reducer(states: BearingStates<any>) {
        const reducer: any = {};
        for (const key in states) {
            if (states.hasOwnProperty(key)) {
                const state = states[key];
                if (state instanceof BearingState) {
                    reducer[key] = state.reducer;
                } else {
                    reducer[key] = this.reducer(state);
                }
            }
        }
        return combineReducers(reducer);
    }


    createStates<T extends BreaingStartOption>(startType: T, key?: string): BearingStates<T> {
        const states: any = {};
        const currentKey = key || '';
        for (const key in startType) {
            if (startType.hasOwnProperty(key)) {
                const state = startType[key];
                if (typeof state === 'object') {
                    states[key] = this.createStates(state, currentKey + '.' + key);
                } else {
                    states[key] = new (state as any)(currentKey + '.' + key);
                }
            }
        }
        return states;
    }

    foreachStates(bstate: BearingStates<any>, callback: (state: BearingState<any>) => void) {
        for (const key in bstate) {
            const state = bstate[key];
            if (state instanceof BearingState) {
                callback(state);
            } else {
                this.foreachStates(state as BearingStates<any>, callback);
            }
        }
    }
}

