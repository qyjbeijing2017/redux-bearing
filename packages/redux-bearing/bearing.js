"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.combinedReducer2SamePoint = void 0;
const redux_1 = require("redux");
const bearing_state_1 = require("./bearing-state");
const redux_devtools_extension_1 = require("redux-devtools-extension");
const combinedReducer2SamePoint = (reducers) => {
    return (state, action) => {
        let stateNew = state;
        for (const reducer of reducers) {
            if (reducer)
                stateNew = Object.assign(Object.assign({}, stateNew), reducer(stateNew, action));
        }
        return stateNew;
    };
};
exports.combinedReducer2SamePoint = combinedReducer2SamePoint;
class Bearing {
    constructor(startType, options) {
        var _a;
        this.release = () => { this.foreachStates(this._states, state => state.release()); };
        this._states = this.createStates(startType);
        const myReducer = this.createReducer(this._states);
        this._store = (0, redux_1.createStore)((0, exports.combinedReducer2SamePoint)([myReducer, options === null || options === void 0 ? void 0 : options.reducer]), (_a = options === null || options === void 0 ? void 0 : options.enhancer) !== null && _a !== void 0 ? _a : (0, redux_devtools_extension_1.composeWithDevTools)());
        this.foreachStates(this._states, (state) => state.store = this._store);
    }
    get store() {
        return this._store;
    }
    get states() {
        return this._states;
    }
    get storeStates() {
        return this.store.getState();
    }
    createReducer(states) {
        const reducer = {};
        for (const key in states) {
            if (states.hasOwnProperty(key)) {
                const state = states[key];
                if (state instanceof bearing_state_1.BearingState) {
                    reducer[key] = state.reducer;
                }
                else {
                    reducer[key] = this.createReducer(state);
                }
            }
        }
        return (0, redux_1.combineReducers)(reducer);
    }
    createStates(startType, key) {
        const states = {};
        const currentKey = key || '';
        for (const key in startType) {
            if (startType.hasOwnProperty(key)) {
                const state = startType[key];
                if (typeof state === 'object') {
                    states[key] = this.createStates(state, currentKey + '.' + key);
                }
                else {
                    states[key] = new state(currentKey + '.' + key);
                }
            }
        }
        return states;
    }
    foreachStates(bstate, callback) {
        for (const key in bstate) {
            const state = bstate[key];
            if (state instanceof bearing_state_1.BearingState) {
                callback(state);
            }
            else {
                this.foreachStates(state, callback);
            }
        }
    }
}
exports.default = Bearing;
//# sourceMappingURL=bearing.js.map