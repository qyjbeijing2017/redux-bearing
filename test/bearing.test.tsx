import { states } from './bearing';
import renderer, { ReactTestRendererJSON } from 'react-test-renderer';
import { BearingComponent } from './bearing-component';
describe('redux-bearing', () => {
    test('create bearing', () => {
        expect(states.num.val).toBe(0);
        expect(states.str.val).toBe('test bearing');
        expect(states.sub.num.val).toBe(1);
        expect(states.sub.str.val).toBe('sub bearing');
    });

    test('update bearing', () => {
        const listener = jest.fn((val) => val);

        states.num.on('update', listener);
        states.num.val = 3;

        expect(states.num.val).toBe(3);
        expect(listener.mock.calls).toHaveLength(1);
        expect(listener.mock.calls[0][0]).toBe(3);
    });

    // it worked in the browser, but not in the test
    it('renders correctly', () => {
        states.num.val = 0;
        const component = renderer.create(<BearingComponent /> as any);
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();

        renderer.act(() => {
            if(!tree) throw new Error('tree is null');
            (tree as ReactTestRendererJSON).props.onClick();
        });
        tree = component.toJSON();
        expect(tree).toMatchSnapshot();

        renderer.act(() => {
            if(!tree) throw new Error('tree is null');
            (tree as ReactTestRendererJSON).props.onClick();
        });
        tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});