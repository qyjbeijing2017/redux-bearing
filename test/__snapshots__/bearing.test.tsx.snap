// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`redux-bearing renders correctly 1`] = `
<button
  onClick={[Function]}
>
  0
</button>
`;

exports[`redux-bearing renders correctly 2`] = `
<button
  onClick={[Function]}
>
  0
</button>
`;

exports[`redux-bearing renders correctly 3`] = `
<button
  onClick={[Function]}
>
  0
</button>
`;
